documentation/specs/src/economics/fee-system.md

```diff
@@ -1,7 +1,376 @@
+is initiated.
+
+From the consensus block proposer's address (included in the Tendermint
+request), it is possible to derive the relative Namada address for the payment:
+should, for any reason, the proposer's address be missing in the incoming
+request, fees for that block will be burned.
+
+The `Fee` field of `WrapperTx` is defined as follows:
+
+```rust
+pub struct Fee {
+  /// amount of the fee
+  pub amount: Amount,
+  /// address of the token
+  pub token: Address,
+}
+```
+
+The signer of the wrapper transaction defines the token in which fees must be
+paid among those available in the token whitelist. At the same time, he also
+sets the amount which must meet the minimum price per gas unit for that token,
+$GP_{min}$ (also defined in the whitelist). The difference between the minimum
+and the actual value set by the submitter represents the incentive for the block
+proposer to prefer the inclusion of this transaction over other ones.
+
+The block proposer can check the validity of these two parameters while
+constructing the block. These validity checks are also replicated in
+`process_proposal` and `mempool_check`. In case a transaction with invalid
+parameters ended up in a block, the entire block would be rejected (as already
+explained earlier in this document). As mentioned before, a signer might submit
+more than one transaction per block and the proposer should take into
+consideration the updated value of the unshielded balance.
+
+Since the whitelist can be changed via governance, transactions could fail these
+checks in the block where the whitelist change happens. For `mempool_check`, the
+checks could reject transactions that may become valid in the future or
+vice-versa: since we can assume a slow rate of change for these parameters and
+mempool and block space optimizations are a priority, it is up to the clients to
+track any changes in these parameters and act accordingly.
+
+### Unshielding
+
+To provide improved privay, Namada allows the signer of the wrapper transaction
+to unshield some funds on the go to cover the cost of the fee. This also
+addresses a possible locked-out problem in which a user doesn't have enough
+funds to pay fees (preventing any sort of operation on the chain). The
+`WrapperTx` struct must be extended as follows:
+
+```rust
+pub struct WrapperTx {
+  /// The fee to be paid for including the tx
+  pub fee: Fee,
+  /// Used to determine an implicit account of the fee payer
+  pub pk: common::PublicKey,
+  /// Max amount of gas that can be used when executing the inner tx
+  pub gas_limit: GasLimit,
+  /// The optional unshielding transaction for fee payment
+  pub unshield: Option<Transaction>,
+  /// the encrypted payload
+  pub inner_tx: EncryptedTx,
+  /// sha-2 hash of the inner transaction acting as a commitment
+  /// the contents of the encrypted payload
+  pub tx_hash: Hash,
+}
+```
+
+The new `unshield` field carries an optional masp transaction struct. The
```

Comment:
what is the type of a "MASP transaction struct"? presumably we should use that in the field above